jdbc/src/main/java/com/interface21/jdbc/mapper/Mapper.java

```diff
@@ -0,0 +1,97 @@
+package com.interface21.jdbc.mapper;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.IntStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Mapper {
+
+    private static final Logger log = LoggerFactory.getLogger(Mapper.class);
+
+    private Mapper() {
+    }
+
+    public static <T> List<T> doQueryMapping(Class<T> clazz, String sqlStatement, ResultSet resultSet)
+            throws SQLException {
+        if (resultSet.isAfterLast()) {
+            return null;
+        }
+
+        List<String> fieldNames = getFieldNameFrom(sqlStatement);
+        final List<T> queryResult = new ArrayList<>();
+
+        while (resultSet.next()) {
+            T newInstance = createInstanceFromResultSet(clazz, resultSet, fieldNames);
+            queryResult.add(newInstance);
+        }
```

Comment:
그렇네요.

하지만 그건 API사용하는 유저입장에서 고려해야할 일이라고 생각합니다.
예를 들어 지금의 `findById` 쿼리는 `jdbcTemplate.query(User.class, sql, id);`처럼 클래스, sql문, id 세개를 매개변수로 받고 있습니다.
`select account from users where id = 1`과 같은 쿼리를 날릴려면 `String account`를 필드로 갖고있는 `AccountMapper`과 같은 데이터 매핑 클래스를 만들어서 `jdbcTemplate.query(AccountMapper.class, sql, id);`를 날려야합니다.

UserDao에서는 항상 User만을 조회할테니 그 부분까지는 고려를 못했습니다.
엔티티의 전부가 아닌 부분 정보를 쿼리할 때 `Map`을 반환하는 것도 하나의 방법이 될 수 있겠네요.
JPA처럼 영속성 컨텍스트를 관리해볼까 생각했었는데 조금 더 template의 본연 기능 집중해야겠습니다. 

```
This class simplifies the use of JDBC and helps to avoid common errors. 
It executes core JDBC workflow, leaving application code to provide SQL and extract results. 
This class executes SQL queries or updates, initiating iteration over ResultSets and catching JDBC exceptions and translating them to the common org.springframework.dao exception hierarchy.
```

[공식 문서](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/jdbc/core/JdbcTemplate.html)에는 요렇게 되어있네요!

Reply:
ResultSet에서 받아오는 것으로 변경해서 해결되었네요! 👍